File: backend/app/schemas/insights.py
"""Wire contracts for the insights endpoints.

The service-layer dataclasses use Python ``date`` and ``Decimal``;
these Pydantic models pin the JSON shape clients see. Decimals
serialise as strings on the wire so JS clients don't lose precision
to float coercion — same convention as the expenses surface.
"""

from __future__ import annotations

from datetime import date
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, ConfigDict

from app.models.enums import ExpenseCategory


class CategoryTotalOut(BaseModel):
    """One row of the monthly breakdown."""

    model_config = ConfigDict(from_attributes=True)

    category: ExpenseCategory
    total: Decimal
    count: int
    average: Decimal


class MonthlyBreakdownOut(BaseModel):
    """Response for ``GET /insights/monthly``."""

    model_config = ConfigDict(from_attributes=True)

    month: date
    grand_total: Decimal
    grand_count: int
    items: list[CategoryTotalOut]


class TrendBucketOut(BaseModel):
    """One ``(month, category)`` cell. ``month`` is the first day of
    that month — clients render that as the bucket label."""

    model_config = ConfigDict(from_attributes=True)

    month: date
    category: ExpenseCategory
    total: Decimal


class CategoryTrendsOut(BaseModel):
    """Response for ``GET /insights/trends``.

    The ``months`` and ``categories`` arrays let a chart library
    pre-allocate axes; ``buckets`` is the dense grid in row-major
    order (month-by-month, then category-by-category). A front-end
    that wants a Recharts-style series array can group buckets by
    ``category`` in one pass.
    """

    model_config = ConfigDict(from_attributes=True)

    months: list[date]
    categories: list[ExpenseCategory]
    buckets: list[TrendBucketOut]


class AnomalyOut(BaseModel):
    """One row of the anomaly response."""

    model_config = ConfigDict(from_attributes=True)

    expense_id: UUID
    merchant_name: str
    category: ExpenseCategory
    amount: Decimal
    expense_date: date
    z_score: float
    baseline_mean: Decimal
    baseline_stddev: Decimal
    baseline_samples: int


class AnomalyReportOut(BaseModel):
    """Response for ``GET /insights/anomalies``.

    The window dates and threshold are echoed back to the client so
    the UI can render "we looked at the last 30 days against the
    previous 6 months" without re-doing the date math.
    """

    model_config = ConfigDict(from_attributes=True)

    lookback_start: date
    baseline_start: date
    z_threshold: float
    anomalies: list[AnomalyOut]

File: frontend/src/api/insights.ts
import { useQuery } from '@tanstack/react-query';

import type { ExpenseCategory } from './expenses';
import { apiFetch } from './client';

// Wire types mirror ``backend/app/schemas/insights.py`` and
// ``schemas/budget.py``. Decimals serialise as strings on the wire
// (same convention as expenses) — keep them as strings here and
// coerce at the render boundary so JS floats don't eat precision
// before the user sees the number.

// ----- monthly breakdown ---------------------------------------------------

export interface CategoryTotal {
  category: ExpenseCategory;
  total: string;
  count: number;
  average: string;
}

export interface MonthlyBreakdown {
  month: string; // ISO date — first of the month
  grand_total: string;
  grand_count: number;
  items: CategoryTotal[];
}

export function useMonthlyBreakdown(month?: string) {
  // ``month`` is the optional query param. We pass it through as-is —
  // the backend accepts any date inside the target month.
  const search = month ? `?month=${encodeURIComponent(month)}` : '';
  return useQuery<MonthlyBreakdown>({
    queryKey: ['insights', 'monthly', month ?? 'today'],
    queryFn: () => apiFetch<MonthlyBreakdown>(`/insights/monthly${search}`),
  });
}

// ----- trends --------------------------------------------------------------

export interface TrendBucket {
  month: string;
  category: ExpenseCategory;
  total: string;
}

export interface CategoryTrends {
  months: string[];
  categories: ExpenseCategory[];
  buckets: TrendBucket[];
}

export function useCategoryTrends(months: number = 12) {
  return useQuery<CategoryTrends>({
    queryKey: ['insights', 'trends', months],
    queryFn: () => apiFetch<CategoryTrends>(`/insights/trends?months=${months}`),
  });
}

// Pivot the dense (month × category) bucket grid into the row-per-
// month shape Recharts wants for stacked bars. One row looks like::
//
//   { month: "2026-04", food_dining: "12.50", groceries: "84.20", ... }
//
// Categories with no spend in any month are dropped at the page
// layer so the chart legend stays readable.
export function pivotTrendsForChart(trends: CategoryTrends): Array<Record<string, string>> {
  const byMonth: Record<string, Record<string, string>> = {};
  for (const month of trends.months) {
    byMonth[month] = { month };
  }
  for (const bucket of trends.buckets) {
    // ``byMonth[bucket.month]`` is guaranteed populated by the loop
    // above — the dense grid invariant means every bucket's month
    // is in ``trends.months``.
    byMonth[bucket.month]![bucket.category] = bucket.total;
  }
  return trends.months.map((m) => byMonth[m]!);
}

// ----- anomalies ----------------------------------------------------------

export interface Anomaly {
  expense_id: string;
  merchant_name: string;
  category: ExpenseCategory;
  amount: string;
  expense_date: string;
  z_score: number;
  baseline_mean: string;
  baseline_stddev: string;
  baseline_samples: number;
}

export interface AnomalyReport {
  lookback_start: string;
  baseline_start: string;
  z_threshold: number;
  anomalies: Anomaly[];
}

export function useAnomalies() {
  return useQuery<AnomalyReport>({
    queryKey: ['insights', 'anomalies'],
    queryFn: () => apiFetch<AnomalyReport>('/insights/anomalies'),
  });
}

// ----- budget status -------------------------------------------------------

export type BudgetPeriod = 'monthly';

export interface BudgetStatusItem {
  budget_id: string;
  category: ExpenseCategory;
  period: BudgetPeriod;
  amount: string;
  spent: string;
  remaining: string;
  ratio: number;
  alert_threshold_pct: number;
  alert_triggered: boolean;
  period_start: string;
  period_end: string;
}

export interface BudgetStatusReport {
  today: string;
  items: BudgetStatusItem[];
}

export function useBudgetStatus() {
  return useQuery<BudgetStatusReport>({
    queryKey: ['budgets', 'status'],
    queryFn: () => apiFetch<BudgetStatusReport>('/budgets/status'),
  });
}
